python/lsst/ctrl/oods/msgIngester.py

```diff
@@ -0,0 +1,167 @@
+        self.msgQueue = MsgQueue(brokers, group_id, topics, self.max_messages)
+
+        butler_configs = self.config["butlers"]
+        if len(butler_configs) == 0:
+            raise Exception("No Butlers configured; check configuration file")
+
+        self.butlers = []
+        for butler_config in butler_configs:
+            butler = ButlerProxy(butler_config["butler"], csc)
+            self.butlers.append(butler)
+
+        self.tasks = []
+        self.dequeue_task = None
+
+    def get_butler_clean_tasks(self):
+        """Get a list of all butler run_task methods
+
+        Returns
+        -------
+        tasks: `list`
+            A list containing each butler run_task method
+        """
+        tasks = []
+        for butler in self.butlers:
+            tasks.append(butler.clean_task)
+        return tasks
+
+    async def ingest(self, butler_file_list):
+        """Attempt to perform butler ingest for all butlers
+
+        Parameters
+        ----------
+        butler_file_list: `list`
+            files to ingest
+        """
+
+        # for each butler, attempt to ingest the requested file,
+        # Success or failure is noted in a message description which
+        # will send out via a CSC logevent.
+        try:
+            for butler in self.butlers:
+                await butler.ingest(butler_file_list)
+        except Exception as e:
+            LOGGER.warn("Exception: %s", e)
+
+    def run_tasks(self):
+        """run tasks to queue files and ingest them"""
+
+        # this is split into two tasks so they can run at slightly different
+        # cadences.  We want to gather as many files as we can before we
+        # do the ingest
+        task = asyncio.create_task(self.msgQueue.queue_files())
+        self.tasks.append(task)
+
+        task = asyncio.create_task(self.dequeue_and_ingest_files())
+        self.tasks.append(task)
+
+        clean_tasks = self.get_butler_clean_tasks()
+        for clean_task in clean_tasks:
+            task = asyncio.create_task(clean_task())
+            self.tasks.append(task)
+
+        return self.tasks
+
+    def stop_tasks(self):
+        self.running = False
+        self.msgQueue.stop()
+        for task in self.tasks:
+            task.cancel()
+        self.tasks = []
+
+    async def dequeue_and_ingest_files(self):
+        self.running = True
+        while self.running:
+            message_list = await self.msgQueue.dequeue_messages()
+            resources = []
+            for m in message_list:
+                rps = self._gather_all_resource_paths(m)
+                resources.extend(rps)
+            await self.ingest(resources)
+            # XXX - commit on success, failure, or metadata_failure
+            for m in message_list:
+                self.msgQueue.commit(message=m)
+
+    def _gather_all_resource_paths(self, m):
+        # extract all urls within this message
+        msg = BucketMessage(m.value())
+
+        rp_list = list()
```

Comment:
Can't this be a (faster) one-liner: `rp_list = [ResourcePath(url) for url in msg.extract_urls()]` ?

Reply:
I'm not sure it's faster, but I changed it.